### PR TITLE
Allow user to select tables from database.

### PR DESCRIPTION
**Pull request by lajones**
Partial fix for #2751. Sets up the code to allow users to select tables and tests.

Still to add: what the command line looks like.

### CODE REVIEW

File: test/EntityFramework.SqlServer.Design.FunctionalTests/ReverseEngineering/SqlServerE2ETests.cs [hunk 1]
```diff
@@ -29,8 +47,8 @@ public SqlServerE2ETests(SqlServerE2EFixture fixture, ITestOutputHelper output)
         }
 
         protected override E2ECompiler GetCompiler() => new E2ECompiler
-        {
-            NamedReferences =
+            {
+                NamedReferences =
                     {
```
Comment (lajones) [inline test/EntityFramework.SqlServer.Design.FunctionalTests/ReverseEngineering/SqlServerE2ETests.cs]: I'll fix these whitespace changes. Not sure how they crept in.

File: src/EntityFramework.Relational.Design/ReverseEngineering/RelationalMetadataModelProvider.cs [hunk 2]
```diff
@@ -50,10 +50,16 @@ public abstract class RelationalMetadataModelProvider : IDatabaseMetadataModelPr
             ModelUtilities = modelUtilities;
         }
 
-        public virtual IModel GenerateMetadataModel([NotNull] string connectionString)
+        public virtual IModel GenerateMetadataModel(
+            [NotNull] string connectionString, [CanBeNull] TableSelectionSet tableSelectionSet)
         {
             Check.NotEmpty(connectionString, nameof(connectionString));
 
+            if (tableSelectionSet != null)
```
Comment (natemcmaster) [inline src/EntityFramework.Relational.Design/ReverseEngineering/RelationalMetadataModelProvider.cs]: Nit: `_tableSelectionSet = tableSelectionSet ?? TableSelectionSet.InclusiveAll;` would be clearer.

File: src/EntityFramework.Relational.Design/ReverseEngineering/TableSelection.cs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    {
+        public const string Any = "*";
+        public static readonly TableSelection InclusiveAll = new TableSelection();
+        public static readonly TableSelection ExclusiveAll = new TableSelection() { Exclude = true };
+
+        public virtual string Schema { get;[param: NotNull] set; } = Any;
+        public virtual string Table { get;[param: NotNull] set; } = Any;
+        public virtual bool Exclude { get; set; }
+
+        public override bool Equals(object o)
```
Comment (natemcmaster) [inline src/EntityFramework.Relational.Design/ReverseEngineering/TableSelection.cs]: Do we need this?

File: src/EntityFramework.Relational.Design/ReverseEngineering/TableSelection.cs [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+            {
+                return false;
+            }
+
+            return Exclude == tableSelection.Exclude
+                && Schema == tableSelection.Schema
+                && Table == tableSelection.Table;
+        }
+
+        public override int GetHashCode()
```
Comment (natemcmaster) [inline src/EntityFramework.Relational.Design/ReverseEngineering/TableSelection.cs]: Same

**Comment by natemcmaster** [comment]
Need more tests. There is some fairly complex logic about how inclusive and exclusive lists are build. Add unit tests to ensure this behaves as expected.

File: src/EntityFramework.Relational.Design/ReverseEngineering/RelationalMetadataModelProvider.cs (hunk 2 above)
Comment (lajones) [inline src/EntityFramework.Relational.Design/ReverseEngineering/RelationalMetadataModelProvider.cs]: Will do.

File: src/EntityFramework.Relational.Design/ReverseEngineering/TableSelection.cs (hunk 3 above)
Comment (lajones) [inline src/EntityFramework.Relational.Design/ReverseEngineering/TableSelection.cs]: I did have an equals comparison at one stage - but I think it got removed. I'll check.

File: src/EntityFramework.Relational.Design/ReverseEngineering/TableSelection.cs (hunk 4 above)
Comment (lajones) [inline src/EntityFramework.Relational.Design/ReverseEngineering/TableSelection.cs]: Required if Equals is required. I'll check.

**Comment by lajones** [comment]
Agree on the tests - I'll add.

File: src/EntityFramework.Relational.Design/ReverseEngineering/RelationalMetadataModelProvider.cs (hunk 2 above)
Comment (lajones) [inline src/EntityFramework.Relational.Design/ReverseEngineering/RelationalMetadataModelProvider.cs]: Done.

File: src/EntityFramework.Relational.Design/ReverseEngineering/TableSelection.cs (hunk 4 above)
Comment (lajones) [inline src/EntityFramework.Relational.Design/ReverseEngineering/TableSelection.cs]: No longer needed. Removed.

File: src/EntityFramework.Relational.Design/ReverseEngineering/TableSelection.cs (hunk 3 above)
Comment (lajones) [inline src/EntityFramework.Relational.Design/ReverseEngineering/TableSelection.cs]: No longer needed. Removed.

**Comment by lajones** [comment]
@natemcmaster - new commit - contains fixes for all issues mentioned before.

File: src/EntityFramework.Relational.Design/ReverseEngineering/TableSelectionSet.cs [hunk 5]
```diff
@@ -0,0 +1,106 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using JetBrains.Annotations;
+using Microsoft.Data.Entity.Utilities;
+
+namespace Microsoft.Data.Entity.Relational.Design.ReverseEngineering
```
Comment (natemcmaster) [inline src/EntityFramework.Relational.Design/ReverseEngineering/TableSelectionSet.cs]: Internal?

File: src/EntityFramework.Relational.Design/ReverseEngineering/TableSelection.cs [hunk 6]
```diff
@@ -0,0 +1,28 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using JetBrains.Annotations;
+using Microsoft.Data.Entity.Utilities;
+
+namespace Microsoft.Data.Entity.Relational.Design.ReverseEngineering
```
Comment (natemcmaster) [inline src/EntityFramework.Relational.Design/ReverseEngineering/TableSelection.cs]: Also Internal?

**Comment by natemcmaster** [comment]
The tests look good :+1: In the spirit of #2956, should we be adding more to the public API? Or can we internalize some of this? /cc: @divega ?

**Comment by lajones** [comment]
@natemcmaster @divega The purpose of having them public is to allow a future UI to construct `TableSelection`s and hence a `TableSelectionSet` and pass it in directly.

**Comment by natemcmaster** [comment]
Right. I mean in the .Internal namespace, but still "public class"

**Comment by lajones** [comment]
@divega? Maybe I misunderstand what the Internal namespace is meant for. I thought it was for classes which we depend on internally but which we do _not_ recommend other people use. Here I would be happy for other code to use this. However, it is true that at the moment no other such code exists and usage by a UI is a way off.

**Comment by divega** [comment]
@lajones #2956 says that anything that isn't consumed by provider should be moved to the internal namespace, but I can see why we may not want to follow that in this case. I would like to defer until we do the API review, which should be soon anyway.

**Comment by lajones** [comment]
@natemcmaster @divega OK - so can I go ahead and check in as it stands?

**Comment by divega** [comment]
Actually, let's make these internal for beta8. We can always bring them back in rc1.

**Comment by lajones** [comment]
OK. Will do. Then will check in.

**Comment by lajones** [comment]
Checked in with commit d3c4001.

**Pull request closed without merging**